subcommand/server-acl-init/command.go

```diff
@@ -664,7 +756,9 @@ func (c *Command) configureConnectInject(logger hclog.Logger, consulClient *api.
 	if err != nil {
 		return err
 	}
-	if len(existingRules) > 0 {
+	// If Consul namespaces are enabled, someone may have changed their namespace
+	// configuration, so auth methods and binding rules should be updated accordingly
+	if len(existingRules) > 0 && !c.flagEnableNamespaces {
```

Comment:
This is actually resulting in a new binding rule being created each time if a non-default namespace is used:
```
consul acl binding-rule list -namespace luke
25c500b4-433f-78f1-9a1f-a73e23e3c49b:
   Namespace:    luke
   AuthMethod:   consul-k8s-auth-method
   Description:  Kubernetes binding rule
   BindType:     service
   BindName:     ${serviceaccount.name}
   Selector:     serviceaccount.name!=default
479039a1-cfe5-61db-6bde-45a922ee272c:
   Namespace:    luke
   AuthMethod:   consul-k8s-auth-method
   Description:  Kubernetes binding rule
   BindType:     service
   BindName:     ${serviceaccount.name}
   Selector:     serviceaccount.name!=default
8e68f7db-a25f-6c71-76b3-013c74394834:
   Namespace:    luke
   AuthMethod:   consul-k8s-auth-method
   Description:  Kubernetes binding rule
   BindType:     service
   BindName:     ${serviceaccount.name}
   Selector:     serviceaccount.name!=default
bf01cb84-8a80-d7ca-d8a6-74c39cff642a:
   Namespace:    luke
   AuthMethod:   consul-k8s-auth-method
   Description:  Kubernetes binding rule
   BindType:     service
   BindName:     ${serviceaccount.name}
   Selector:     serviceaccount.name!=default
```

This is because the call to list the rules isn't namespaced and so `existingRules` is `[]`. Then when we iterate through it to check if the rule already exists, it never does and so we create a new one.

I think the logic needs to be: if namespaces are not enabled, get the existing rules, check if they're empty, if they are then exit early. Else, continue on and when we get to the code below where the existingRules slice was previously used, instead make another call with the correct namespace.

Reply:
Fixed, as well as the issue that a user's config change to binding rules would never update the rule.